### ISSUE TITLE
Provide Command Line Interface for Ensemble Copula Coupling

### ISSUE THREAD

**Issue by gavinevans** [metoppv/improver#56]
As a post-processing suite trialler, I want to be able to run ECC using a script, which can be called from the command line so that Ensemble Copula Coupling can run and create output for verification.

Related issues: #30, #5 

**Acceptance criteria:**
 * Ensure that a Command Line Interface exists for ECC. This should allow all of following processes:
   1. Generate a set of percentiles from another set of percentiles and then reorder to create members.
   ~2. Generate percentiles from probabilities and then reorder to create members.~
   3. Generate a set of percentiles from another set of percentiles with no reordering. Rebadge as members.
   ~4. Generate percentiles from probabilities with no reordering. Rebadge as members.~
 * A CLI test should be added, as appropriate, see #5. This should  have acceptance tests for  1 & 2, and maybe one of 3 &4 from the processes above.
* The CLI should be able to output a different number of members to those input, by using the relevant conversion plugins (there is one for interpolating percentiles and one for interpolating probabilities to percentiles)
* Only write to disk once during the CLI (i.e. don't output intermediate steps)
* Keep interface as simple as possible.

**Information**
* The reordering can be done in two ways, either based on the original ensemble members or in a random way and both should be supported.
* Spot data is out of scope at the moment, but tests for this will be covered by follow on tickets.  #48

**Note**
After this ticket was started the support of probabilistic data was made out of scope, which has now been added to #95

**Comment by gavinevans** [comment]
The interfaces to the plugins are:

- `RebadgePercentilesAsMembers`
```python
process(cube, ensemble_member_numbers=None)
```
- `ResamplePercentiles`
```python
process(self, forecast_at_percentiles, no_of_percentiles=None,
             sampling="quantile")
```
- `GeneratePercentilesFromProbabilities`
```python
process(self, forecast_probabilities, no_of_percentiles=None,
             sampling="quantile")
```
- `GeneratePercentilesFromMeanAndVariance`
```python
process(self, calibrated_forecast_predictor_and_variance,
             no_of_percentiles)
```
- `EnsembleReordering`
```python
process(self, post_processed_forecast, raw_forecast,
             random_ordering=False)
```

**Comment by fionaRust** [comment]
Update from Jonathan, so modifying the acceptance criteria and updating the information on this ticket before it is re-refined later on.

**Comment by gavinevans** [comment]
A cube with multiple realizations has been saved as follows. This is to help produce a cube with percentiles and probabilities.

```python
# Save an example cube containing realizations. 
# Taken from suite mi-aq649_laea for cycletime 20170217T0300Z.
cubes = iris.load("./enukx_stage_m*_t12/level2a/*.nc", "wind_speed")
for cube in cubes:
    cube.coord("ensemble_member_id").rename("realization")
    cube.attributes.pop("history")

cubelist = iris.cube.CubeList([])
for cube in cubes:
    cube = iris.util.new_axis(cube, "realization")
    cubelist.append(cube)

cube = cubelist.concatenate_cube()

constr = iris.Constraint(height=10)
multi_height_level_cube = cube.copy()
cube = cube.extract(constr)

iris.save(cube, "./multiple_realization_wind_cube.nc")
iris.save(multi_height_level_cube, "./multiple_realization_and_heights_wind_cube.nc")
```

**Comment by gavinevans** [comment]
A percentile cube has been created as follows, using the existing `GeneratePercentilesFromMeanAndVariance` plugin:
```python
# Create percentiles from members.
import iris

from improver.ensemble_copula_coupling.ensemble_copula_coupling import (
    GeneratePercentilesFromMeanAndVariance)

cube = iris.load_cube("./multiple_realization_wind_cube.nc")

#Mean
ensemble_mean = cube.collapsed('realization', iris.analysis.MEAN)
#Variance
ensemble_std_dev = cube.collapsed('realization', iris.analysis.STD_DEV)

percentile_cube = GeneratePercentilesFromMeanAndVariance().process(
    [ensemble_mean, ensemble_std_dev], 12)

iris.save(percentile_cube, "multiple_percentiles_wind_cube.nc")
```

**Comment by gavinevans** [comment]
**Branches related to this work**
**Branch for work on ECC CLI**
Need to decide on the best option for using `argparse` when sub-options are required for CLI options.
https://github.com/gavinevans/improver/tree/ecc_cli

**Branch to fix bug in dictionary**
The key of the dictionary needs to be made useable by both the `GeneratePercentilesFromProbabilities` plugin and the `ResamplePercentiles` plugin. The key of the dictionary should be the name of the phenomenon.
https://github.com/gavinevans/improver/tree/ecc_dictionary_key_fix

**Branch to refactor cubes that contain percentiles/probabilities to fit with temporary conventions**
https://github.com/gavinevans/improver/tree/ecc_probabilistic_refactor

**Comment by gavinevans** [comment]
As described in https://github.com/metoppv/improver/issues/110#issuecomment-310660261, the issue regarding which approach to use for `argparse` has been unblocked. For the moment, incompatible sub-options will be handled by raising an exception. 

This will be investigated further in #139. Some relevant discussion is available in https://realpython.com/blog/python/comparing-python-command-line-parsing-libraries-argparse-docopt-click/.

**Comment by gavinevans** [comment]
Some sample wind speed data, which has been used so far for developing CLI acceptance tests is available from @fionaRust. These files may need checking for correspondence with the current temporary conventions for probabilistic data.

**Comment by gavinevans** [comment]
Note that the changes to `ensemble_copula_coupling.py` and `ensemble_copula_coupling_constants.py` within https://github.com/gavinevans/improver/tree/ecc_cli will likely need reverting, if the decision is made to progress with the adoption of the temporary conventions for probabilistic data witihin https://github.com/gavinevans/improver/tree/ecc_probabilistic_refactor.

**Comment by arh89** [comment]
Assuming that we're going forward with the temporary conventions (else I see no reason we can progress), I've rebased Gavin's ecc_probabilistic_refactor and ecc_cli branches on top of master (resolving conflicts such that the ecc_probabilistic_refactor changes in `ensemble_copula_coupling.py` and `ensemble_compula_coupling_constants.py` were taken from the ecc_probabilistic_refactor branch...

The new branch (which should be worked from), is here:
https://github.com/arh89/improver/tree/ecc_cli

**Comment by gavinevans** [comment]
@arh89 @fionaRust 
For @benfitzpatrick, how many of these possible CLI tests for the different options should we actually be including within our CLI tests? 

Possible options for CLI tests are:

- ResamplePercentiles(quantile sampling method) + EnsembleReordering(default ordering) - Default
- ResamplePercentiles(random sampling method) + EnsembleReordering(default ordering)
- ResamplePercentiles(quantile sampling method) + EnsembleReordering(random ordering)
- ResamplePercentiles(random sampling method) + EnsembleReordering(random ordering)
- ResamplePercentiles(quantile sampling method) + RebadgePercentilesAsMembers(default numbering) - Default
- ResamplePercentiles(random sampling method) + RebadgePercentilesAsMembers(default numbering)
- ResamplePercentiles(quantile sampling method) + RebadgePercentilesAsMembers(apply renumbering)
- ResamplePercentiles(random sampling method) + RebadgePercentilesAsMembers(apply renumbering)

**Comment by benfitzpatrick** [comment]
One reordering and one rebadging are good enough for me

**Comment by arh89** [comment]
@gavinevans and @benfitzpatrick, it is probably worth noting that the current (development) version of the CLI tests will resample the percentiles to produce a number which is equal to the number of ensemble members (12), rather than to the previous the number of percentiles (9, based on the input data) - the default behaviour (if `--no_of_percentiles` is not passed in).

Does this mean that the default behaviour is not the most common use case? If not, should it be?

**Comment by fionaRust** [comment]
I have plotted the output from the CLI tests. These rebadged output looked ok, but the resordered test looked a bit odd.

I regenerated the test data with 18 percentiles to start with, rather than 9. This produced results that look a lot more reasonable. I think this would be a better acceptance test to do

I have some ipython notebooks with the results in if anyone want to look.

**Comment by gavinevans** [comment]
@fionaRust I'm happy to use the test data with 18 percentiles as the known good output, as it contains notably fewer artefacts than the test data with 9 percentiles. Thanks!

**Comment by fionaRust** [comment]
It should be noted that the scope of this issue was reduced to only cover percentile data as input. Input of probabilities will have to be handled in another issue later on.

**Comment by gavinevans** [comment]
Added slide to sprint demo.

**Comment by fionaRust** [comment]
The CLI support for pecentiles was complete in pull request #158 
The threhold support was not done as part of this issue, and has been added to the user story for ECC support for verification. #95
